apps/passport-server/src/services/generic-issuance/pipelines/PretixPipeline.ts

```diff
@@ -7,37 +52,723 @@ import {
+  externalId: string; // Pretix's event ID
+  genericIssuanceId: string; // Our UUID
+  name: string; // Display name for the event
+  products: PretixProductConfig[];
+}
+
+/**
+ * Configuration for specific products available for the event. Does not need
+ * to include all products available in Pretix, but any product listed here
+ * must be available in Pretix.
+ */
+export interface PretixProductConfig {
+  externalId: string; // Pretix's item ID
+  genericIssuanceId: string; // Our UUID
+  name: string; // Display name
+  isSuperUser: boolean; // Is a user with this product a "superuser"?
+  // Superusers are able to check tickets in to events.
+}
+
+/**
+ * Class encapsulating the complete set of behaviors that a {@link Pipeline} which
+ * loads data from Pretix is capable of.
  */
 export class PretixPipeline implements BasePipeline {
   public type = PipelineType.Pretix;
   public capabilities = [
-    // TODO: fill this out with an issuance and checkin capability
+    {
+      issue: this.issuePretixTicketPCDs.bind(this),
+      feedId: "ticket-feed",
+      type: PipelineCapability.FeedIssuance,
+      getFeedUrl: (): string => generateIssuanceUrlPath(this.id)
+    } satisfies FeedIssuanceCapability,
+    {
+      checkin: this.checkinPretixTicketPCDs.bind(this),
+      type: PipelineCapability.Checkin,
+      getCheckinUrl: (): string => generateCheckinUrlPath(this.id)
+    } satisfies CheckinCapability
   ];
 
+  /**
+   * Used to sign {@link EdDSATicketPCD}
+   */
   private eddsaPrivateKey: string;
   private definition: PretixPipelineDefinition;
-  private db: IPipelineAtomDB;
+
+  /**
+   * This is where the Pipeline stores atoms so that they don't all have
+   * to be stored in-memory.
+   */
+  private db: IPipelineAtomDB<PretixAtom>;
+  private api: IGenericPretixAPI;
 
   public get id(): string {
     return this.definition.id;
   }
 
+  public get issuanceCapability(): FeedIssuanceCapability {
+    return this.capabilities[0] as FeedIssuanceCapability;
+  }
+
+  public get checkinCapability(): CheckinCapability {
+    return this.capabilities[1] as CheckinCapability;
+  }
+
   public constructor(
     eddsaPrivateKey: string,
     definition: PretixPipelineDefinition,
-    db: IPipelineAtomDB
+    db: IPipelineAtomDB,
+    api: IGenericPretixAPI
   ) {
     this.eddsaPrivateKey = eddsaPrivateKey;
     this.definition = definition;
-    this.db = db;
+    this.db = db as IPipelineAtomDB<PretixAtom>;
+    this.api = api;
   }
 
+  /**
+   * Loads external data from Lemonade and saves it to the {@link IPipelineAtomDB} for
+   * later use.
+   *
+   * TODO:
+   * - clear tickets after each load? important!!!!
+   */
   public async load(): Promise<void> {
+    logger(LOG_TAG, `loading for pipeline id ${this.id}`);
+    const tickets: PretixTicket[] = [];
+
+    const errors = [];
+
+    for (const event of this.definition.options.events) {
+      // @todo this can throw exceptions. how should we handle this?
```

Comment:
one idea is to aggregate errors warnings, and save them somewhere that can be loaded + displayed to a user.